lib/db/transpose-and-aggregate-by-dimensions.js

```diff
@@ -0,0 +1,83 @@
+const _ = require('lodash')
+
+module.exports = function (params) {
+  const { data, metric, dimensions, timeDimension } = params
+
+  const { transpositionDimension, aggregationDimension } = parseDimensions(dimensions, timeDimension)
+
+  const transposedData = transposeAndAggregate({ data, metric, transpositionDimension, aggregationDimension })
+
+  const placeholders = createZeroValuePlaceholdersForTranpose(transposedData)
+
+  return mergePlaceholders(transposedData, placeholders)
+}
+
+function parseDimensions (dimensions, timeDimension) {
+  const [transpositionDimension, aggregationDimension] = dimensions.map(dimension => {
+    if (dimension === timeDimension) {
+      return 'timestamp'
+    }
+    return dimension
+  })
+
+  return { transpositionDimension, aggregationDimension }
+}
+
+function transposeAndAggregate (params) {
+  const {
+    data,
+    metric,
+    transpositionDimension,
+    aggregationDimension
+  } = params
+  return _.chain(data)
+    .reduce((store, record) => {
+      const transposedRecord = transposeRecord({ record, metric, aggregationDimension, transpositionDimension })
+      if (!transposedRecord) return store
+      return updateStore({ store, aggregationDimension, record: transposedRecord })
+    }, {})
+    .values()
+    .value()
+}
+
+function transposeRecord ({ record, metric, aggregationDimension, transpositionDimension }) {
+  if (!record[transpositionDimension] && record[transpositionDimension] !== '') return
+
+  const transposedKey = record[transpositionDimension] === '' ? 'emptyString' : record[transpositionDimension]
+
+  return {
+    [aggregationDimension]: record[aggregationDimension],
+    [transposedKey]: Number(record[metric])
+  }
+}
+
+function updateStore ({ store, aggregationDimension, record }) {
+  const aggregationKey = record[aggregationDimension]
+  const currentValue = _.get(store, [`${aggregationKey}`], {})
+  _.set(store, [`${aggregationKey}`], { ...currentValue, ...record })
```

Comment:
do you need to use string interpolation here?

Reply:
Ah, thanks for catching this.  It looks like I can remove the string interpolation as long as I leave the square brackets;  all of this was originally added because I tested with aggregationKey set to strings that included `.`, e.g. `hug.arcgis.com` and that caused problems with _.get and _.set.  Square brackets are enough to pass the tests though!  I'll remove string interpolation and commit.